src/container/streamAnalyticsManager.ts

```diff
@@ -4,40 +4,129 @@
-import { AzureAccount } from "../typings/azure-account.api";
+import { AzureAccount, AzureSession, AzureSubscription } from "../typings/azure-account.api";
 import { StreamAnalyticsPickItem } from "./models/streamAnalyticsPickItem";
 
+export enum ASAUpdateStatus {
+    Idle,
+    CheckingUpdate,
+    Updating,
+}
+
 export class StreamAnalyticsManager {
+    public static getInstance(): StreamAnalyticsManager {
+        if (!StreamAnalyticsManager.instance) {
+            StreamAnalyticsManager.instance = new StreamAnalyticsManager();
+            return StreamAnalyticsManager.instance;
+        }
+        return StreamAnalyticsManager.instance;
+    }
+
+    private static instance: StreamAnalyticsManager;
     private readonly azureAccount: AzureAccount;
     private readonly MaximumRetryCount: number = 3;
+    private asaUpdateStatus: ASAUpdateStatus;
 
-    constructor() {
+    private constructor() {
         this.azureAccount = vscode.extensions.getExtension<AzureAccount>("ms-vscode.azure-account")!.exports;
+        this.asaUpdateStatus = ASAUpdateStatus.Idle;
     }
 
-    public async selectStreamingJob(): Promise<StreamAnalyticsPickItem> {
+    public async getJobInfo(): Promise<any> {
         await Utility.waitForAzLogin(this.azureAccount);
-
-        const job: StreamAnalyticsPickItem = await vscode.window.showQuickPick(this.loadAllStreamingJobs(), { placeHolder: `Select ${Constants.asaJobDesc}`, ignoreFocusOut: true });
-        if (!job) {
+        const streamingJob: StreamAnalyticsPickItem = await vscode.window.showQuickPick(this.loadAllStreamingJobs(), { placeHolder: `Select ${Constants.asaJobDesc}`, ignoreFocusOut: true });
+        if (!streamingJob) {
             throw new UserCancelledError();
         }
-        return job;
+
+        return await vscode.window.withProgress({
+            location: vscode.ProgressLocation.Notification,
+            title: "Querying Stream Analytics Job information",
+        }, async (): Promise<object> =>  {
+            const ASAJobResourceId: string = streamingJob.job.id;
+            return await this.queryASAJobInfo(ASAJobResourceId, streamingJob.azureSubscription.session);
+        });
     }
 
-    public async getJobInfo(streamingJob: StreamAnalyticsPickItem): Promise<object> {
-        try {
-            const id: string = streamingJob.job.id;
-            const publishJobUrl: string = `https://management.azure.com${id}/publishedgepackage?api-version=2017-04-01-preview`;
+    public async checkAndUpdateASAJob(templateFile: string, moduleName: string) {
+        if (this.asaUpdateStatus === ASAUpdateStatus.Idle) {
+            await Utility.waitForAzLogin(this.azureAccount);
+            try {
+                this.asaUpdateStatus = ASAUpdateStatus.CheckingUpdate;
+                const isUpdateAvailable = await this.isASAJobUpdateAvailable(templateFile, moduleName);
+                this.asaUpdateStatus = ASAUpdateStatus.Idle;
+                if (isUpdateAvailable) {
```

Comment:
Seems check/update are in two lifecycle. The edge case could be, user click check -> but not aware of the update and do not take action. Then the ASAJob is updated again. And user check update again -> and user aware of the previous notification, however it update the value to the previous update, not the latest one. I am thinking whether the message box should be a modal dialog to ask for user input directly.

Reply:
Until user click update button, then we will try to query latest changes and update the module twin and it would be latest one.